README.md
# rs-mysql cookbook

[![Release](https://img.shields.io/github/release/rightscale-cookbooks/rs-mysql.svg?style=flat)][release]
[![Build Status](https://img.shields.io/travis/rightscale-cookbooks/rs-mysql.svg?style=flat)][travis]

[release]: https://github.com/rightscale-cookbooks/rs-mysql/releases/latest
[travis]: https://travis-ci.org/rightscale-cookbooks/rs-mysql

Provides recipes for managing a MySQL server with RightScale, including:

* Automatic tuning based on available system memory and server usage type
* High availability with master/slave replication including failover
* Database storage with cloud volumes including backup and restore

Github Repository: [https://github.com/rightscale-cookbooks/rs-mysql](https://github.com/rightscale-cookbooks/rs-mysql)

# Requirements

* Requires Chef 12
* Requires Ruby 2.3.1 of higher
* [RightLink 10](http://docs.rightscale.com/rl10/)
* See cookbook version 1.2.6 for chef11 support
* See cookbook version 1.1.9 for RightLink 6 support
* Platform
  * Ubuntu 12.04(MySQL 5.5, MySQL 5.6)
  * Ubuntu 14.04(MySQL 5.5, MySQL 5.6)
  * CentOS 6(MySQL 5.5, MySQL 5.6, MySQL 5.7)
  * CentOS 7(MySQL 5.5, MySQL 5.6, MySQL 5.7)
* Cookbooks
  * [marker](http://community.opscode.com/cookbooks/marker)
  * [mysql](http://community.opscode.com/cookbooks/mysql)
  * [collectd](https://github.com/rightscale-cookbooks-contrib/chef-collectd)
  * [database](https://github.com/rightscale-cookbooks-contrib/database)
  * [rightscale_tag](http://community.opscode.com/cookbooks/rightscale_tag)
  * [filesystem](http://community.opscode.com/cookbooks/filesystem)
  * [lvm](http://community.opscode.com/cookbooks/lvm)
  * [rightscale_volume](http://community.opscode.com/cookbooks/rightscale_volume)
  * [rightscale_backup](http://community.opscode.com/cookbooks/rightscale_backup)
  * [chef_handler](http://community.opscode.com/cookbooks/chef_handler)
  * [dns](https://github.com/rightscale-cookbooks-contrib/dns)

See the `Berksfile` and `metadata.rb` for up to date dependency information.

# Usage

To setup a standalone MySQL server, place the `rs-mysql::default` recipe in the runlist.

## Creating a new volume

To create a new volume and move the MySQL database to it, run the `rs-mysql::volume` recipe with the following
attributes set:

- `node['rs-mysql']['device']['nickname']` - the nickname of the volume
- `node['rs-mysql']['device']['volume_size']` - the size of the volume to create
- `node['rs-mysql']['device']['filesystem']` - the filesystem to use on the volume
- `node['rs-mysql']['device']['mount_point']` - the location to mount the volume

This will create a new volume, attach it to the server, format it with the filesystem specified, mount it on the
location specified, and move the MySQL database to it.

### Provisioned IOPS on EC2

To create a volume with IOPS on EC2, set the following attribute before running the `rs-mysql::volume` recipe:

- `node['rs-mysql']['device']['iops']` - the value of IOPS to use

## Creating a logical volume with striping

To create a logical volume with striping using LVM and move the MySQL database to it, run the `rs-mysql::stripe` recipe
with the following attributes set:

- `node['rs-mysql']['device']['nickname']` - the nickname to use as prefix for the logical volume
- `node['rs-mysql']['device']['count']` - number of volumes to create in the logical volume
- `node['rs-mysql']['device']['volume_size']` - the total size of the logical volume
- `node['rs-mysql']['device']['filesystem']` - the filesystem to use on the logical volume
- `node['rs-mysql']['device']['mount_point']` - the location to mount the logical volume

This will create the number of volumes specified in `node['rs-mysql']['device']['count']`. Each volume created will have
a nickname of `"#{nickname}_#{device_number}"`. The size for each volume is calculated by the following formula:

```ruby
(total_size.to_f / device_count.to_f).ceil

# For example, total size = 10, stripe count = 3
(10.0 / 3.0).ceil
# => 4.0
```

This will create a volume group with the name `"#{nickname}-vg"` and a logical volume in it with the name
`"#{nickname}-lv"`, format it with the filesystem specified, mount it on the location specified, and move the MySQL
database to it.

## Master/Slave replication and Failover

To setup a MySQL master/slave replication, place the `rs-mysql::master` recipe in the runlist for the master
server and the `rs-mysql::slave` recipe in the runlist for the slave server. The master server should be
operational before bringing up the slave server. Both the master and slave servers are tagged with required
information for replication. Please refer to the [rightscale_tag] cookbook for more information about the tags
added to database servers. When using volumes with master/slave replication, the `rs-mysql::volume` or
`rs-mysql::stripe` recipe should run before the `rs-mysql::master` or `rs-mysql::slave` recipe.

To promote a slave to master in a failover situation just run the `rs-mysql::master` recipe on the slave that needs to
be promoted. If there are other slaves in the deployment, the `rs-mysql::slave` recipe should be re-run on those servers
as well. If the old master is still working, it can be demoted to a slave by running the `rs-mysql::slave` recipe as
well.

## Backing up volume(s) & Cleaning up backups

To create a backup of all volumes attached to the server, run the `rs-mysql::backup` recipe with the following
attributes set:

- `node['rs-mysql']['backup']['lineage']` - the lineage to be used for the backup

The backup process will create a snapshot of all volumes attached to the server (except the boot disk if there is one).
During the backup process, the MySQL database will be read-only for a period to ensure a consistent backup, so it is
recommended that it only runs on a slave MySQL database server. The backup recipe also handles the cleanup of old
volume snapshots and accepts the following attributes:

- `node['rs-mysql']['backup']['keep']['keep_last']` - number of last backups to keep from deleting
- `node['rs-mysql']['backup']['keep']['dailies']` - number of daily backups to keep
- `node['rs-mysql']['backup']['keep']['weeklies']` - number of weekly backups to keep
- `node['rs-mysql']['backup']['keep']['monthlies']` - number of monthly backups to keep
- `node['rs-mysql']['backup']['keep']['yearlies']` - number of yearly backups to keep

This will cleanup the old snapshots on the cloud based on the criteria specified.

## Restoring a volume from a backup

To restore a volume from backup, run the `rs-mysql::volume` recipe with the same set of attributes mentioned in the
[previous section](#creating-a-new-volume) along with the following attribute:

- `node['rs-mysql']['restore']['lineage']` - the lineage to restore the backup from

This will restore the volume from the backup instead of creating a new one. By default, the backup with the latest
timestamp will be restored. To restore backup from a specific timestamp, set the following attribute:

- `node['rs-mysql']['restore']['timestamp']` - the timestamp of the backup to restore from (in seconds since UNIX epoch)

## Restoring a logical volume composed of multiple volumes from a backup

To restore a logical volume composed of multiple volumes from a backup, run the `rs-mysql::stripe` recipe with the
same set of attributes mentioned in the [previous section](#creating-stripe-of-volumes) along with the following
attribute:

- `node['rs-mysql']['restore']['lineage']` - the lineage to restore the backup from

This will restore multiple volumes from the backup matching the lineage. By default, the backup with the latest
timestamp will be restored. To restore a backup from a specific timestamp, set the following attribute:

- `node['rs-mysql']['restore']['timestamp']` - the timestamp of the backup to restore from (in seconds since UNIX epoch)

## Scheduling automated backups of volume(s)

To schedule automated backups, run the `rs-mysql::schedule` recipe with the following attributes set:

- `node['rs-mysql']['schedule']['enable']` - to enable/disable automated backups
- `node['rs-mysql']['schedule']['hour']` - the hour to take the backup on (should use crontab syntax)
- `node['rs-mysql']['schedule']['minute']` - the minute to take the backup on (should use crontab syntax)
- `node['rs-mysql']['backup']['lineage']` - the lineage name to be used for the backup

This will create a crontab entry to run the `rs-mysql::backup` recipe periodically at the given minute and hour. To
disable the automated backups, simply set `node['rs-mysql']['schedule']['enable']` to `false` and rerun the
`rs-mysql::schedule` recipe and this will remove the crontab entry.

## Deleting volume(s)

This operation should be part of the decommission bundle in a RightScale ServerTemplate where the volumes attached to
the server are detached and deleted from the cloud but this can also be used as an operational recipe. This recipe will
do nothing in the following conditions:

- when the server enters a stop state
- when the server reboots

This recipe also has a safety attribute `node['rs-mysql']['device']['destroy_on_decommission']`. This attribute will be
set to `false` by default and should be overridden and set to `true` in order for the devices to be detached and
deleted. If an LVM is found (created using `rs-mysql::stripe`), the LVM will be conditionally removed before detaching
the volume.

# Attributes

- `node['rs-mysql']['server_usage']` - The server usage type. It should be either `'dedicated'` or `'shared'`.
  Default is `'dedicated'`.
- `node['rs-mysql']['server_root_password']` - The root password for MySQL server.
- `node['rs-mysql']['server_repl_password']` - The replication password for MySQL server.
- `node['rs-mysql']['application_username']` - The database username to be created for the application.
- `node['rs-mysql']['application_password']` - The database password to be used for the application user.
- `node['rs-mysql']['application_user_privileges']` - The application user's privileges.
- `node['rs-mysql']['application_database_name']` - The name of the application database.
- `node['rs-mysql']['import']['repository']` - Repository location containing the database dump file to import.
- `node['rs-mysql']['import']['private_key']` - The private key to access the repository via SSH.
- `node['rs-mysql']['import']['revision']` - The revision of the database dump file to import.
- `node['rs-mysql']['import']['dump_file']` - Filename of the database dump file to import.
- `node['rs-mysql']['dns']['master_fqdn']` - The fully qualified domain name of the master database.
- `node['rs-mysql']['dns']['user_key']` - The user key for the DNS provider to access/modify DNS
records.
- `node['rs-mysql']['dns']['secret_key']`- The secret key for the DNS provider to access/modify DNS
records.
- `node['rs-mysql']['device']['nickname']` - The nickname of the device or the logical volume comprising multiple
  devices. Default is `'data_storage'`.
- `node['rs-mysql']['device']['mount_point']` - The mount point for the device. Default is `'/mnt/storage'`.
- `node['rs-mysql']['device']['volume_size']` - The size (in gigabytes) of the volume to be created. If multiple
  devices are used, this will be the total size of the logical volume. Default is `10`.
- `node['rs-mysql']['device']['count']` - The number of devices to be created for the logical volume. Default is `2`.
- `node['rs-mysql']['device']['iops']` - The IOPS value to be used for EC2 Provisioned IOPS. This attribute should only
  be used with Amazon EC2. Default is `nil`.
- `node['rs-mysql']['device']['filesystem']` - The filesystem to be used on the device. Default is `'ext4'`.
- `node['rs-mysql']['device']['detach_timeout']` - Amount of time (in seconds) to wait for a volume to detach at
  decommission. Default is `300` (5 minutes).
- `node['rs-mysql']['device']['destroy_on_decommission']` - Whether to destroy the device during the decommission of the
  server. Default is `false`.
- `node['rs-mysql']['device']['mkfs_options']` - Additional mkfs options for formatting the device. Default is `'-F'`
  which is required to avoid warnings about formatting the whole device.
- `node['rs-mysql']['device']['stripe_size']` - The stripe size to use on LVM. Default is `512`.
- `node['rs-mysql']['backup']['lineage']` - The backup lineage. Default is `nil`.
- `node['rs-mysql']['backup']['keep']['keep_last']` - Maximum snapshots to keep. Default is `60`.
- `node['rs-mysql']['backup']['keep']['dailies']` - Number of daily backups to keep. Default is `14`.
- `node['rs-mysql']['backup']['keep']['weeklies']` - Number of weekly backups to keep. Default is `6`.
- `node['rs-mysql']['backup']['keep']['monthlies']` - Number of monthly backups to keep. Default is `12`.
- `node['rs-mysql']['backup']['keep']['yearlies']` - Number of yearly backups to keep. Default is `2`.
- `node['rs-mysql']['restore']['lineage']` - The name of the lineage to restore the backups from. Default is `nil`.
- `node['rs-mysql']['restore']['timestamp']` - The timestamp to restore backup taken on or before the timestamp in the
  same lineage. Default is `nil`.
- `node['rs-mysql']['schedule']['enable']` - Enable/disable automated backups. Default is `false`.
- `node['rs-mysql']['schedule']['hour']` - The backup schedule hour. Default is `nil`.
- `node['rs-mysql']['schedule']['minute']` - The backup schedule minute. Default is `nil`.
- `node['rightscale']['decom_reason']` - Set from RL10 Shutdown Reason Rightscript to determine how to handle
   rs-mysql::decommission behavior

# Recipes

## `rs-mysql::default`

Installs the MySQL server and tunes the attributes used in the `my.cnf` based on the available system memory
and the server usage type. If the server usage type is `'dedicated'`, all resources in the system are dedicated
to the MySQL server and if the usage type is `'shared'`, only half of the resources are used for the MySQL server.
This `'shared'` usage will be used in building a LAMP stack where the same system is used to run both the MySQL
server and the PHP application server. This recipe also tags the server as a standalone MySQL server.

## `rs-mysql::volume`

Creates a new volume from scratch or from an existing backup based on the value provided in the
`node['rs-mysql']['restore']['lineage']` attribute. If this attribute is set, the volume will be restored from a
backup matching this lineage otherwise a new volume will be created from scratch. This recipe will also format the
volume using the filesystem specified in `node['rs-mysql']['device']['filesystem']`, mount the volume on the location
specified in `node['rs-mysql']['device']['mount_point']`, and move the MySQL database directory to the volume.

## `rs-mysql::stripe`

Creates a new logical volume composed of volumes from scratch or from an existing backup based on the
value provided in the `node['rs-mysql']['restore']['lineage']` attribute. If this attribute is set, the volumes will be
restored from a backup matching this lineage otherwise a new logical volume composed of volumes will be created from
scratch. This recipe will create a striped logical volume using LVM on the volumes and format the logical volume
using the filesystem specified in `node['rs-mysql']['device']['filesystem']`. This will also mount the volume on the
location specified in `node['rs-mysql']['device']['mount_point']` and move the MySQL database directory to the logical
volume.

## `rs-mysql::master`

This recipe sets up the database to act as the master. It makes sure the database is not read only by overriding
the `mysql/tunable/read_only` to false and includes the `rs-mysql::default` recipe which installs MySQL and
performs the configuration. The master database specific tags are added to the server and the master is reset.
The master database can be provided with a fully qualified domain name (FQDN) by setting the
`node['rs-mysql']['dns']['master_fqdn']` attribute. The DNS provider credentials
(`node['rs-mysql']['dns']['user_key']` and `node['rs-mysql']['dns']['secret_key']`) must also be set
to create/update the DNS records in the DNS provider.

## `rs-mysql::slave`

This recipe modifies the MySQL server to be read only and includes the `rs-mysql::default` recipe which installs
MySQL, performs configuration, and tags the server as a slave server. It obtains the information about the master
database with the help of the [`find_database_servers`] helper method provided by the [rightscale_tag] cookbook and
changes the master host of the slave to the latest master available in the deployment. If this recipe is run after
`rs-mysql::volume` or `rs-mysql::stripe` and a backup was restored, this recipe will use information from the backup to
assist with catching up with the master MySQL database.

## `rs-mysql::backup`

Takes a backup of all volumes attached to the server (except boot disks if there are any) with the lineage specified
in the `node['rs-mysql']['backup']['lineage']` attribute. During the backup process, the MySQL database will be
read-only and the filesystem will be frozen. The filesystem will be unfrozen and the MySQL database will no longer be
read-only after the backup even if the backup process fails with the help of a chef exception handler. This recipe will
also clean up the volume snapshots based on the criteria specified in the `rs-mysql/backup/keep/*` attributes.

## `rs-mysql::schedule`

Adds or removes the crontab entry for taking backups periodically at the minute and hour provided via
`node['rs-mysql']['schedule']['minute']` and `node['rs-mysql']['schedule']['hour']` attributes. The recipe uses the
`node['rs-mysql']['schedule']['enable']` attribute to determine whether to add or remove the crontab entry.

## `rs-mysql::decommission`

If the `node['rs-mysql']['device']['destroy_on_decommission']` attribute is set to true, this recipe moves the MySQL
database back to the root volume, drops the database specified by `node['rs-mysql']['application_database_name']` if it
is specified, and detaches and deletes the volumes attached to the server. This operation will be skipped if the server
is entering the stop state or rebooting.

## `rs-mysql::dump_import`

Retrieves a dump file from a Git repository and imports it to the database server. The Git repository is
specified by `node['rs-mysql']['import']['repository']` with revision/branch specified by
`node['rs-mysql']['import']['revision']`. The dump file in the repository is specified by
`node['rs-mysql']['import']['dump_file']`. The private key attribute, `node['rs-mysql']['import']['private_key']`,
must be set if retrieving from a private repository.

## `rs-mysql::collectd`

Installs the collectd plugins for MySQL.

[rightscale_tag]: https://github.com/rightscale-cookbooks/rightscale_tag/blob/master/README.md
[`find_database_servers`]: https://github.com/rightscale-cookbooks/rightscale_tag#find_database_servers

# Author

Author:: RightScale, Inc. (<cookbooks@rightscale.com>)


CHANGELOG.md
rs-mysql Cookbook CHANGELOG
=======================

This file is used to list changes made in each version of the rs-mysql cookbook.

v2.0.3
------
- updates filesystem cookbook to >= 0.11.1 to use the supermarket version and \
  fix the frozen? issue.
- cleans up Berksfile

v2.0.2
------
- adds ephemeral_lvm > 3.0
- allows customizable mysql server versions, refer to README.md for supported versions.
- updated test-kitchen to allow for testing for multiple versions.

v2.0.1
------
- adds azure rm support for private ips, issue: [#68][]

v2.0.0
------
- Remove support for Chef 11
- Add support for Chef 12

v1.2.7
------
- pins nio4r

v1.2.6
------
- fixing selinux permission error when moving directories on centos.
- updating rake task to check for changelog, and metadata version

v1.2.5
------
- Pinning cookbooks and gems for ruby and chef compatibility

v1.2.4
------

- Adds ruby slave monitors

v1.2.3
------

- Updates collectd monitors to work with collectd v5

v1.2.2
------

- Run pvscan after volumes attached for stripe.  Behavior seen in CentOS 6.6 where volumes attached are not
  immediately seen as part of a Logical Volume.

v1.2.1
------

- Hard Setting versions in metadata and berksfile
- Locking down aws, ohai, and build-essential so it doesn't pull in chef12 libs

v1.2.0
------
- add support for RightLink 10

v1.1.9
------

- No longer use IP address to create server_id since some clouds use the same IP but forward using ports.

v1.1.8
------

- Updated lvm cookbook dependency version to 1.3.6.
- Set default filesystem on RHEL 7 platform_family to xfs.

v1.1.7
------

- On RHEL, depending on cloud, check and wait for RHEL repos to be installed.

v1.1.6
------

- Remove logic to ignore cloud/public_ips for server-id on cloudstack.

v1.1.5
------

- Ignore cloud/public_ips for server-id on cloudstack.

v1.1.4
------

- Use Upstart for the mysql service in decommission on Ubuntu 14.04.

v1.1.3
------

- Fix volume type input description since it is not just for vSphere.

v1.1.2
------

- Add kitchen testing for RHEL 6.5 and Ubuntu 14.04.
- Update versions of dependent cookbooks.
- Update testing code to abide by Serverspec v2.

v1.1.1
------

- Update the dependency versions of `rightscale_volume` and `rightscale_backup`, supporing VMware vSphere.
- Set `rs-mysql/device/volume_type` node attribute in metadata.
- Pass `rs-mysql/device/volume_type` and `rs-mysql/device/controller_type` to `rightscale_volume`.

v1.1.0
------

- Release to coincide with the v14.0.0 Beta ServerTemplates.

v1.0.0
------

- Initial release


metadata.rb
# frozen_string_literal: true
name             'rs-mysql'
maintainer       'RightScale, Inc.'
maintainer_email 'cookbooks@rightscale.com'
license          'Apache 2.0'
description      'Installs and configures a MySQL server'
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version          '2.0.3'
issues_url       'https://github.com/rightscale-cookbooks/rs-mysql/issues' if respond_to?(:issues_url)
source_url       'https://github.com/rightscale-cookbooks/rs-mysql' if respond_to?(:source_url)
chef_version     '>= 12.0' if respond_to?(:chef_version)

depends 'chef_handler'
depends 'marker'
depends 'database'
depends 'mysql2_chef_gem'
depends 'mysql'
depends 'yum-mysql-community'
depends 'collectd'
depends 'rightscale_tag'
depends 'filesystem', '>= 0.11.1'
depends 'ephemeral_lvm', '>= 3.0'
depends 'lvm'
depends 'rightscale_volume'
depends 'rightscale_backup'
depends 'build-essential'
depends 'rs-base', '>= 2.0'
depends 'dns'

recipe 'rs-mysql::default', 'Sets up a standalone MySQL server'
recipe 'rs-mysql::collectd', 'Sets up collectd monitoring for MySQL server'
recipe 'rs-mysql::master', 'Sets up a MySQL master server'
recipe 'rs-mysql::slave', 'Sets up a MySQL slave server'
recipe 'rs-mysql::volume', 'Creates a volume, attaches it to the server, and moves the MySQL data to the volume'
recipe 'rs-mysql::stripe', 'Creates volumes, attaches them to the server, sets up a striped LVM, and moves the MySQL' \
                           ' data to the volume'
recipe 'rs-mysql::backup', description: 'Creates a backup', thread: 'storage_backup'
recipe 'rs-mysql::decommission', 'Destroys LVM conditionally, detaches and destroys volumes. This recipe should' \
                                 ' be used as a decommission recipe in a RightScale ServerTemplate.'
recipe 'rs-mysql::schedule', 'Enable/disable periodic backups based on rs-mysql/schedule/enable'
recipe 'rs-mysql::slave_monitors', 'enables slave scripts'

attribute 'rs-mysql/server_usage',
  display_name: 'Server Usage',
  description: "The Server Usage method. It is either 'dedicated' or 'shared'. In a 'dedicated' server all" \
                  " server resources are dedicated to MySQL. In a 'shared' server, MySQL utilizes only half of the resources." \
                  " Example: 'dedicated'",
  default: 'dedicated',
  required: 'optional',
  recipes: ['rs-mysql::default', 'rs-mysql::master', 'rs-mysql::slave']

attribute 'rs-mysql/bind_network_interface',
  display_name: 'MySQL Bind Network Interface',
  description: 'The network interface to use for MySQL bind. It can be either' \
                  " 'private' or 'public' interface.",
  default: 'private',
  choice: %w(public private),
  required: 'optional',
  recipes: ['rs-mysql::default', 'rs-mysql::master', 'rs-mysql::slave']

attribute 'rs-mysql/server_root_password',
  display_name: 'MySQL Root Password',
  description: 'The root password for MySQL server. Example: cred:MYSQL_ROOT_PASSWORD',
  required: 'required',
  recipes: ['rs-mysql::default', 'rs-mysql::master', 'rs-mysql::slave']

attribute 'rs-mysql/application_username',
  display_name: 'MySQL Application Username',
  description: 'The username of the application user. Example: cred:MYSQL_APPLICATION_USERNAME',
  required: 'optional',
  recipes: ['rs-mysql::default', 'rs-mysql::master', 'rs-mysql::slave']

attribute 'rs-mysql/application_password',
  display_name: 'MySQL Application Password',
  description: 'The password of the application user. Example: cred:MYSQL_APPLICATION_PASSWORD',
  required: 'optional',
  recipes: ['rs-mysql::default', 'rs-mysql::master', 'rs-mysql::slave']

attribute 'rs-mysql/application_user_privileges',
  display_name: 'MySQL Application User Privileges',
  description: 'The privileges given to the application user. This can be an array of mysql privilege types.' \
                  ' Example: select, update, insert',
  required: 'optional',
  type: 'array',
  default: [:select, :update, :insert],
  recipes: ['rs-mysql::default', 'rs-mysql::master', 'rs-mysql::slave']

attribute 'rs-mysql/application_database_name',
  display_name: 'MySQL Database Name',
  description: 'The name of the application database. Example: mydb',
  required: 'optional',
  recipes: ['rs-mysql::default', 'rs-mysql::master', 'rs-mysql::slave']

attribute 'rs-mysql/server_repl_password',
  display_name: 'MySQL Slave Replication Password',
  description: 'The replication password set on the master database and used by the slave to authenticate and' \
                  ' connect. If not set, rs-mysql/server_root_password will be used. Example cred:MYSQL_REPLICATION_PASSWORD',
  required: 'optional',
  recipes: ['rs-mysql::default', 'rs-mysql::master', 'rs-mysql::slave']

attribute 'rs-mysql/device/count',
  display_name: 'Device Count',
  description: 'The number of devices to create and use in the Logical Volume. If this value is set to more than' \
                  ' 1, it will create the specified number of devices and create an LVM on the devices.',
  default: '2',
  recipes: ['rs-mysql::stripe', 'rs-mysql::decommission'],
  required: 'recommended'

attribute 'rs-mysql/device/mount_point',
  display_name: 'Device Mount Point',
  description: 'The mount point to mount the device on. Example: /mnt/storage',
  default: '/mnt/storage',
  recipes: ['rs-mysql::volume', 'rs-mysql::stripe', 'rs-mysql::decommission'],
  required: 'recommended'

attribute 'rs-mysql/device/nickname',
  display_name: 'Device Nickname',
  description: 'Nickname for the device. rs-mysql::volume uses this for the filesystem label, which is' \
                  ' restricted to 12 characters.  If longer than 12 characters, the filesystem label will be set to the' \
                  ' first 12 characters. Example: data_storage',
  default: 'data_storage',
  recipes: ['rs-mysql::volume', 'rs-mysql::stripe', 'rs-mysql::decommission'],
  required: 'recommended'

attribute 'rs-mysql/device/volume_size',
  display_name: 'Device Volume Size',
  description: 'Size of the volume or logical volume to create (in GB). Example: 10',
  default: '10',
  recipes: ['rs-mysql::volume', 'rs-mysql::stripe'],
  required: 'recommended'

attribute 'rs-mysql/device/iops',
  display_name: 'Device IOPS',
  description: 'IO Operations Per Second to use for the device. Currently this value is only used on AWS clouds.' \
                  ' Example: 100',
  recipes: ['rs-mysql::volume', 'rs-mysql::stripe'],
  required: 'optional'

attribute 'rs-mysql/device/volume_type',
  display_name: 'Volume Type',
  description: 'Volume Type to use for creating volumes. Example: gp2',
  recipes: ['rs-mysql::volume', 'rs-mysql::stripe'],
  required: 'optional'

attribute 'rs-mysql/device/filesystem',
  display_name: 'Device Filesystem',
  description: 'The filesystem to be used on the device. Defaults are based on OS and determined in' \
                  ' attributes/volume.rb. Example: ext4',
  recipes: ['rs-mysql::volume', 'rs-mysql::stripe'],
  required: 'optional'

attribute 'rs-mysql/device/detach_timeout',
  display_name: 'Detach Timeout',
  description: 'Amount of time (in seconds) to wait for a single volume to detach at decommission. Example: 300',
  default: '300',
  recipes: ['rs-mysql::volume', 'rs-mysql::stripe'],
  required: 'optional'

attribute 'rs-mysql/device/destroy_on_decommission',
  display_name: 'Destroy on Decommission',
  description: 'If set to true, the devices will be destroyed on decommission.',
  default: 'false',
  recipes: ['rs-mysql::decommission'],
  required: 'recommended'

attribute 'rs-mysql/backup/lineage',
  display_name: 'Backup Lineage',
  description: 'The prefix that will be used to name/locate the backup of the MySQL database server.',
  required: 'required',
  recipes: ['rs-mysql::default', 'rs-mysql::master', 'rs-mysql::slave', 'rs-mysql::backup']

attribute 'rs-mysql/restore/lineage',
  display_name: 'Restore Lineage',
  description: 'The lineage name to restore backups. Example: staging',
  recipes: ['rs-mysql::volume', 'rs-mysql::stripe'],
  required: 'recommended'

attribute 'rs-mysql/restore/timestamp',
  display_name: 'Restore Timestamp',
  description: 'The timestamp (in seconds since UNIX epoch) to select a backup to restore from.' \
                  ' The backup selected will have been created on or before this timestamp. Example: 1391473172',
  recipes: ['rs-mysql::volume', 'rs-mysql::stripe'],
  required: 'recommended'

attribute 'rs-mysql/backup/keep/dailies',
  display_name: 'Backup Keep Dailies',
  description: 'Number of daily backups to keep. Example: 14',
  default: '14',
  recipes: ['rs-mysql::backup'],
  required: 'optional'

attribute 'rs-mysql/backup/keep/weeklies',
  display_name: 'Backup Keep Weeklies',
  description: 'Number of weekly backups to keep. Example: 6',
  default: '6',
  recipes: ['rs-mysql::backup'],
  required: 'optional'

attribute 'rs-mysql/backup/keep/monthlies',
  display_name: 'Backup Keep Monthlies',
  description: 'Number of monthly backups to keep. Example: 12',
  default: '12',
  recipes: ['rs-mysql::backup'],
  required: 'optional'

attribute 'rs-mysql/backup/keep/yearlies',
  display_name: 'Backup Keep Yearlies',
  description: 'Number of yearly backups to keep. Example: 2',
  default: '2',
  recipes: ['rs-mysql::backup'],
  required: 'optional'

attribute 'rs-mysql/backup/keep/keep_last',
  display_name: 'Backup Keep Last Snapshots',
  description: 'Number of snapshots to keep. Example: 60',
  default: '60',
  recipes: ['rs-mysql::backup'],
  required: 'optional'

attribute 'rs-mysql/schedule/enable',
  display_name: 'Backup Schedule Enable',
  description: 'Enable or disable periodic backup schedule',
  default: 'false',
  choice: %w(true false),
  recipes: ['rs-mysql::schedule'],
  required: 'recommended'

attribute 'rs-mysql/schedule/hour',
  display_name: 'Backup Schedule Hour',
  description: "The hour to schedule the backup on. This value should abide by crontab syntax. Use '*' for taking" \
                  ' backups every hour. Example: 23',
  recipes: ['rs-mysql::schedule'],
  required: 'required'

attribute 'rs-mysql/schedule/minute',
  display_name: 'Backup Schedule Minute',
  description: 'The minute to schedule the backup on. This value should abide by crontab syntax. Example: 30',
  recipes: ['rs-mysql::schedule'],
  required: 'required'

attribute 'rs-mysql/dns/master_fqdn',
  display_name: 'MySQL Database FQDN',
  description: 'The fully qualified domain name of the MySQL master database server.',
  required: 'optional',
  recipes: ['rs-mysql::master']

attribute 'rs-mysql/dns/user_key',
  display_name: 'DNS User Key',
  description: 'The user key to access/modify the DNS records.',
  required: 'optional',
  recipes: ['rs-mysql::master']

attribute 'rs-mysql/dns/secret_key',
  display_name: 'DNS Secret Key',
  description: 'The secret key to access/modify the DNS records.',
  required: 'optional',
  recipes: ['rs-mysql::master']

attribute 'rs-mysql/import/private_key',
  display_name: 'Import Secret Key',
  description: 'The private key to access the repository via SSH. Example: Cred:DB_IMPORT_KEY',
  required: 'optional',
  recipes: ['rs-mysql::dump_import']

attribute 'rs-mysql/import/repository',
  display_name: 'Import Repository URL',
  description: 'The repository location containing the database dump file to import.' \
                  ' Example: git://example.com/dbfiles/database_dumpfiles.git',
  required: 'optional',
  recipes: ['rs-mysql::dump_import']

attribute 'rs-mysql/import/revision',
  display_name: 'Import Repository Revision',
  description: 'The revision of the database dump file to import.' \
                  ' Example: master',
  required: 'optional',
  recipes: ['rs-mysql::dump_import']

attribute 'rs-mysql/import/dump_file',
  display_name: 'Import Filename',
  description: 'Filename of the database dump file to import.' \
                  ' Example: dumpfile_20140102.gz',
  required: 'optional',
  recipes: ['rs-mysql::dump_import']

attribute 'rs-mysql/startup-timeout',
  display_name: 'MySQL Server Startup Timeout',
  description: 'MySQL Server Startup Timeout',
  required: 'optional',
  default: 300,
  recipes: ['rs-mysql::default']
